### PR TITLE
Refactoring and optimization

### PR DESCRIPTION
**Pull request by mertwole**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -127,181 +44,191 @@ macro_rules! symbol {
...
-    }
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Program {
+    rules: Vec<Rule>,
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Program(pub Vec<Rule>);
+impl Program {
+    pub fn solve(&self) -> KnowledgeBase {
```
Comment (stargazing-dino) [inline src/lib.rs:53]: I like this being a method on Program

File: src/lib.rs [hunk 2]
```diff
@@ -127,181 +44,191 @@ macro_rules! symbol {
...
 
-impl Display for Term {
-    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        match self {
-            Term::Var(v) => write!(f, "{v}"),
-            Term::Sym(s) => write!(f, "{s:?}"),
-        }
-    }
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Program {
```
Comment (stargazing-dino) [inline src/lib.rs:48]: Yes, this looks better as a struct

File: src/lib.rs [hunk 3]
```diff
@@ -127,181 +44,191 @@ macro_rules! symbol {
...
 
 #[cfg(test)]
 mod tests {
-    use alloc::{format, string::ToString as _};
+    use alloc::string::ToString as _;
 
-    use super::*;
+    use super::{parsing::parse, *};
 
     #[test]
```
Comment (stargazing-dino) [inline src/lib.rs:233]: Do these tests need to feature gate on parsing now though?

```
#[cfg(feature = "parsing")]
```

**Review by stargazing-dino — APPROVED** [review]
Nothing major, just curious about the cfg flags for tests.

I'm glad we trimmed so much stuff off. I feel like something purpose built for our use case is perfect.

File: src/lib.rs (hunk 3 above)
Comment (mertwole) [reply src/lib.rs:233]: nope, `test` is like a feature itself that's enabled only when `cargo test` runs

**Pull request merged**